beacon-chain/slasher/params.go

```diff
@@ -0,0 +1,130 @@
+// which validator chunk index a validator index is for, we simply divide
+// the validator index, i, by K.
+func (p *Parameters) validatorChunkIndex(validatorIndex uint64) uint64 {
+	return validatorIndex / p.ValidatorChunkSize
+}
+
+// Given a validator index, and epoch, we compute the exact index
+// into our flat slice on disk which stores K validators' chunks, each
+// chunk of size C. For example, if C = 3 and K = 3, the data we store
+// on disk is a flat slice as follows:
+//
+//     val0     val1     val2
+//      |        |        |
+//   {     }  {     }  {     }
+//  [2, 2, 2, 2, 2, 2, 2, 2, 2]
+//
+// Then, figuring out the exact cell index for epoch 1 for validator 2 is computed
+// with (validatorIndex % K)*C + (epoch % C), which gives us:
+//
+//  (2 % 3)*3 + (1 % 3) =
+//  (2*3) + (1)         =
+//  7
+//
+//     val0     val1     val2
+//      |        |        |
+//   {     }  {     }  {     }
+//  [2, 2, 2, 2, 2, 2, 2, 2, 2]
+//                        |-> epoch 1 for val2
+//
+func (p *Parameters) cellIndex(validatorIndex, epoch uint64) uint64 {
+	validatorChunkOffset := p.validatorOffset(validatorIndex)
+	chunkOffset := p.chunkOffset(epoch)
+	return validatorChunkOffset*p.ChunkSize + chunkOffset
+}
+
+// Computes the start index of a chunk given an epoch.
+func (p *Parameters) chunkOffset(epoch uint64) uint64 {
+	return epoch % p.ChunkSize
+}
+
+// Computes the start index of a validator chunk given a validator index.
+func (p *Parameters) validatorOffset(validatorIndex uint64) uint64 {
+	return validatorIndex % p.ValidatorChunkSize
+}
+
+// Construct a key for our database schema given a validator index and epoch.
+// This calculation gives us a uint that uniquely represents
+// a 2D chunk given a validator index and epoch value.
+// First, we compute the validator chunk index for the validator index,
+// Then, we compute the chunk index for the epoch.
+// If ChunkSize C = 3 and ValidatorChunkSize K = 3, and HistoryLength H = 12,
+// if we are looking for epoch 6 and validator 6, then
+//
+//  validatorChunkIndex = 6 / 3 = 2
+//  chunkIndex = (6 % HistoryLength) / 3 = (6 % 12) / 3 = 2
+//
+// Then we compute how many chunks there are per max span, known as the "width"
+//
+//  width = H / C = 12 / 3 = 4
+//
+// So every span has 4 chunks. Then, we have a disk key calculated by
+//
+//  validatorChunkIndex * width + chunkIndex = 2*4 + 2 = 10
+//
+func (p *Parameters) flatSliceID(validatorIndex, epoch uint64) uint64 {
+	validatorChunkIndex := p.validatorChunkIndex(validatorIndex)
+	chunkIndex := p.chunkIndex(epoch)
+	width := p.HistoryLength / p.ChunkSize
+	return validatorChunkIndex*width + chunkIndex
+}
+
+// Given a validator chunk index, we determine all of the validator
+// indices that will belong in that chunk.
+func (p *Parameters) validatorIndicesInChunk(validatorChunkIdx uint64) []uint64 {
+	validatorIndices := make([]uint64, 0)
+	low := validatorChunkIdx * p.ValidatorChunkSize
```

Comment:
Overflow? Use `Mul64()`?

Reply:
Impossible to overflow because a validator chunk idx is computed as

`validatorChunkIdx = validatorIdx / ValidatorChunkSize`

Then, to go back to validatorIdx, we simply multiply by ValidatorChunkSize